sway/desktop/render.c

```diff
@@ -330,10 +330,35 @@ static void render_saved_view(struct sway_view *view,
 static void render_view(struct sway_output *output, pixman_region32_t *damage,
 		struct sway_container *con, struct border_colors *colors) {
 	struct sway_view *view = con->view;
-	if (!wl_list_empty(&view->saved_buffers)) {
-		render_saved_view(view, output, damage, view->container->alpha);
-	} else if (view->surface) {
-		render_view_toplevels(view, output, damage, view->container->alpha);
+
+	if (!container_is_floating(con)) {
+		struct wlr_box clip_box = {
+			.x = con->current.content_x - output->lx,
+			.y = con->current.content_y - output->ly,
+			.width = con->current.content_width,
+			.height = con->current.content_height,
+		};
+		scale_box(&clip_box, output->wlr_output->scale);
+
+		pixman_region32_t clipped_damage;
+		pixman_region32_init(&clipped_damage);
+		pixman_region32_intersect_rect(&clipped_damage, damage,
+			clip_box.x, clip_box.y, clip_box.width, clip_box.height);
+
+		if (!wl_list_empty(&view->saved_buffers)) {
+			render_saved_view(view, output, &clipped_damage, view->container->alpha);
+		} else if (view->surface) {
+			render_view_toplevels(view, output, &clipped_damage, view->container->alpha);
+		}
+
+		pixman_region32_fini(&clipped_damage);
+
+	} else {
+		if (!wl_list_empty(&view->saved_buffers)) {
+			render_saved_view(view, output, damage, view->container->alpha);
+		} else if (view->surface) {
+			render_view_toplevels(view, output, damage, view->container->alpha);
+		}
```

Comment:
This basically does the same as #5451, but on the other side of the equation.

`render_texture` takes a destination box, which it turns into a pixman rect (using union_rect), and then intersects it with the output damage. #5451 approach limits the destination box size (which is just 2 fmin calls), where this approach limits the output damage (which is an arbitrarily large number of random rectangles).

I think the approach in #5451 lacked handling of output scale, but I wonder if that's all it's missing.

Reply:
Same as above, I did it this way because of simplicity and locality. The intersection call is not free but is the overhead even measurable? If it indeed is, maybe the intersection calls could be moved to an outer function and save us some math operations.

Reply:
> The intersection call is not free but is the overhead even measurable?

All execution time is measurable. It's always a good idea to avoid work in code that is called many times for every frame. That is not to suggest excessive premature optimization, but if it is trivial to pick a *much* cheaper option, then it would make sense to do so.

I am not extremely opposed to your current approach, but I just like the idea of just giving a smaller wlr_box to render_texture to copy instead of adding additional intersects on the output damage.

> If it indeed is, maybe the intersection calls could be moved to an outer function and save us some math operations. In any case, some math has to be done to also account for scale and transformations. GTK AFAIK only uses integer scaling factors, other GUI kits may use floats, so this might be an extra thing to account for, to get the roundings right etc.

This is all accounted for in `render_texture`. Also, wayland only exposes integer scaling factors for buffers.

Reply:
> All execution time is measurable. It's always a good idea to avoid work in code that is called many times for every frame. That is not to suggest excessive premature optimization, but if it is trivial to pick a much cheaper option, then it would make sense to do so.

Agreed. My thinking was that if it works well and no alternative is going to be merged soon, it could be good for now and I would come back and improve it after I fixed some other things I wanted fixed. BTW I just saw that you just updated your PR, going to check it out!